src/java/org/apache/cassandra/hints/HintsDispatcher.java

```diff
@@ -194,38 +307,106 @@ private <T> Action sendHints(Iterator<T> hints, Collection<Callback> callbacks,
+            }
+        }
+        else
+        {
+            // Don't wait for a normal response that will never come since no hints were sent
+            callback.onResponse(null);
+        }
+
         return callback;
     }
 
+    private Pair<Mutation, Hint> splitHintIntoAccordAndNormal(Hint hint)
+    {
+        ClusterMetadata cm = ClusterMetadata.current();
+        Pair<Mutation, Mutation> accordAndNormal = splitMutationIntoAccordAndNormal(hint.mutation, cm);
+        if (accordAndNormal.left == null)
+            return Pair.create(null, hint);
+        if (accordAndNormal.right == null)
+            return Pair.create(accordAndNormal.left, null);
+        Hint normalHint = Hint.create(accordAndNormal.right, hint.creationTime, accordAndNormal.right.smallestGCGS());
+        return Pair.create(accordAndNormal.left, normalHint);
+    }
+
     /*
      * Sending hints in raw mode.
      */
 
     private Callback sendEncodedHint(ByteBuffer hint)
     {
         HintMessage.Encoded message = new HintMessage.Encoded(hostId, hint, messagingVersion);
-        Callback callback = new Callback(message.getHintCreationTime());
+        Callback callback = new Callback(address, message.getHintCreationTime());
         MessagingService.instance().sendWithCallback(Message.out(HINT_REQ, message), address, callback);
         return callback;
     }
 
-    static final class Callback implements RequestCallback
+    static final class Callback implements RequestCallback, Runnable
     {
         enum Outcome { SUCCESS, TIMEOUT, FAILURE, INTERRUPTED }
 
         private final long start = approxTime.now();
         private final Condition condition = newOneTimeCondition();
-        private volatile Outcome outcome;
+        private Outcome normalOutcome;
+        private Outcome accordOutcome;
+        @Nullable
+        private final InetAddressAndPort to;
         private final long hintCreationNanoTime;
+        private final long accordTxnStartNanos;
+        private final Pair<TxnId, Future<TxnResult>> accordTxnResult;
+
+        private Callback(@Nonnull InetAddressAndPort to, long hintCreationTimeMillisSinceEpoch)
+        {
+            this(to, hintCreationTimeMillisSinceEpoch, -1, null);
+        }
 
-        private Callback(long hintCreationTimeMillisSinceEpoch)
+        private Callback(@Nullable InetAddressAndPort to, long hintCreationTimeMillisSinceEpoch, long accordTxnStartNanos, @Nullable Pair<TxnId, Future<TxnResult>> accordTxnResult)
         {
+            this.to = to != null ? to : ACCORD_HINT_ENDPOINT;
             this.hintCreationNanoTime = approxTime.translate().fromMillisSinceEpoch(hintCreationTimeMillisSinceEpoch);
+            this.accordTxnStartNanos = accordTxnStartNanos;
+            this.accordTxnResult = accordTxnResult;
+            if (accordTxnResult != null)
+                accordTxnResult.right.addListener(this, ImmediateExecutor.INSTANCE);
```

Comment:
shouldn't we register as a result/throwable consumer so we are notified if the txn fails unexpectedly?

Reply:
We get the error either way because in `run` it calls `getTxnResult` which does the error handling. Not sure it's worth refactoring this in particular because it makes implementing `getTxnResult` a little more complicated since it would have to work as both a callback and as something invoked on a promise.